lib/src/utils.dart

```diff
@@ -273,6 +273,196 @@ String getLineEnding(String yaml) {
+    /// skipping any white-space. Otherwise, returns the [index] of the next
+    /// non-white-space character.
+    int? skipWhitespace(int index) {
+      var nextIndex = index;
+
+      while (true) {
+        if (nextIndex == yaml.length) return null;
+        if (yaml[nextIndex].trim().isNotEmpty) return nextIndex;
+        ++nextIndex;
+      }
+    }
+
+    var currentOffset = currentEndOffset;
+
+    externalLoop:
+    while (true) {
+      if (currentOffset == yaml.length) break;
+
+      var leadingChar = yaml[currentOffset].trim();
+      var indexOfCommentStart = -1;
+
+      if (leadingChar.isEmpty) {
+        switch (skipWhitespace(currentOffset)) {
+          case final int nextIndex:
+            currentOffset = nextIndex;
+            leadingChar = yaml[currentOffset];
+            break;
+
+          default:
+            currentOffset = yaml.length;
+            break externalLoop; // Exit loop entirely!
+        }
+      }
+
+      /// We need comments only, nothing else. This may be pointless but will
+      /// help us avoid extracting comments when provided random offsets
+      /// within a string.
+      if (leadingChar == '#') indexOfCommentStart = currentOffset;
+
+      /// This is a mindless assumption that the last character was either
+      /// `\n` or [white-space] or the last erroneus offset provided.
+      if (indexOfCommentStart == -1) break;
+
+      final indexOfLineBreak = yaml.indexOf(lineEnding, currentOffset);
+      final isEnd = indexOfLineBreak == -1;
+
+      final comment = yaml
+          .substring(indexOfCommentStart, isEnd ? null : indexOfLineBreak)
+          .trim();
+
+      if (comment.isNotEmpty) comments.add(comment);
+
+      if (isEnd) {
+        currentOffset += comment.length;
+        break;
+      }
+      currentOffset = indexOfLineBreak + 1; // Skip line-break eagerly
+    }
+
+    return (currentOffset, comments);
+  }
+
+  return (
+    nextStartOffset,
+    yaml.substring(currentEndOffset, nextStartOffset).split(lineEnding).fold(
+      <String>[],
+      (buffer, current) {
+        final comment = current.trim();
+        if (comment.isNotEmpty) buffer.add(comment);
+        return buffer;
+      },
+    )
+  );
+}
+
+/// Normalizes an encoded [YamlNode] encoded as a string by pruning any
+/// dangling line-breaks.
+///
+/// This function checks the last `YamlNode` of the [update] that is a
+/// `YamlScalar` and removes any unwanted line-break within the
+/// [updateAsString].
+///
+/// This is achieved by obtaining the chunk of the [yaml] that is after the
+/// current node being replaced using its [nodeToReplaceEndOffset]. If:
+///   1. The chunk has any trailing line-break then the it is left untouched.
+///   2. The node being replaced with [update] is not the last node, then it
+///      is left untouched.
+///   3. The terminal node in [update] is a `YamlScalar`, that is,
+///      the last [YamlNode] within the [update] that is not a collection.
```

Comment:
Please consider adding some example.

Also why is it that we don't do this normalization when we create the `updateAsString`? Wouldn't that be more correct -- I don't know -- I probably don't understand all the context here, that's why I'm asking :D

Reply:
That's the point of adding the dangling line-break. 

`yamlEncodeBlock` has no context itself of which method called it or the current structure of the yaml document. It could be:

1. `updateInList` or `_appendToBlockList` or `_insertInBlockList` for lists
2. `_replaceInBlockMap` or `_addToBlockMap` for block maps
3. `Yaml.update` for top level `YamlNode`.

Thus, the callers should make a call to the function to prune the trailing line break because they have an idea of the yaml's structure.

Additionally, I found it quite advantageous on our part to try and include line breaks for an existing `YamlNode` being replaced to edits. This reduces bugs and removes the need to try and guess where the next line break is. 

It better to include it and check later if it was there 🤷🏽‍♂️